k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java

```diff
@@ -0,0 +1,409 @@
+                }
+                commands.add(builder.build());
+            }
+
+        } else {
+            //This should never happen
+            commands = Collections.singletonList(this);
+        }
+
+        return commands;
+
+    }
+
+    public SearchResponse execute() throws MessagingException {
+
+        try {
+            List<List<ImapResponse>> responses = executeInternal(false);
+            return SearchResponse.parseMultiple(commandFactory, responses);
+        } catch (IOException ioe) {
+            throw folder.ioExceptionHandler(commandFactory.getConnection(), ioe);
+        }
+
+    }
+
+    private void addSequenceNumbers(StringBuilder builder) {
+        addGrouping(builder, sequenceSet, sequenceRanges);
+    }
+
+    private void addUids(StringBuilder builder) {
+        if (uidSet != null || uidRanges != null) {
+            builder.append("UID ");
+            addGrouping(builder, uidSet, uidRanges);
+        }
+    }
+
+    private void addGrouping(StringBuilder builder, Set<Long> numberSet,
+                             List<Map.Entry<String, String>> numberRanges) {
+        optimizeGrouping(numberSet, numberRanges);
+        List<String> sequenceElements = new ArrayList<>();
+        if (numberSet != null) {
+            for (Long number : numberSet) {
+                sequenceElements.add(String.valueOf(number));
+            }
+        }
+        if (numberRanges != null) {
+            for (Map.Entry<String, String> numberRange : numberRanges) {
+                sequenceElements.add(numberRange.getKey() + ":" + numberRange.getValue());
+            }
+        }
+        if (sequenceElements.size() > 0) {
+            builder.append(TextUtils.join(",", sequenceElements.toArray()));
+            builder.append(" ");
+        }
+    }
+
+    private void optimizeGrouping(Set<Long> numberSet, List<Map.Entry<String, String>> numberRanges) {
+        //TODO minimize the number of elements in both the lists
+    }
+
+    private void addQueryString(StringBuilder builder) {
+        //TODO use a literal string instead of a quoted string
+        if (queryString != null) {
+            String encodedQuery = ImapUtility.encodeString(queryString);
+            if (performFullTextSearch) {
+                builder.append("TEXT ");
+            } else {
+                builder.append("OR SUBJECT ").append(encodedQuery).append(" FROM ");
+            }
+            builder.append(encodedQuery).append(" ");
+        }
+    }
+
+    private void addSince(StringBuilder builder) {
+        if (since != null) {
+            SimpleDateFormat rfc3501DateFormat = new SimpleDateFormat("dd-MMM-yyyy", Locale.US);
+            builder.append("SINCE ").append(rfc3501DateFormat.format(since)).append(" ");
+        }
+    }
+
+    private void addRequiredFlags(StringBuilder builder) {
+        if (requiredFlags != null) {
+            for (Flag flag : requiredFlags) {
+                switch (flag) {
```

Comment:
I was thinking of moving the strings into instance variables in the Flag enum as well